### PR TITLE
cli: Add a `jj git init` command.

### PR DESCRIPTION
**Pull request by essiene**
Implement `jj git init`

There are two flags which work as mostly expected.

See `GitInitArgs` for details.

Some improvements around `colocated` repos vs `--git-repo=.` will come after this PR is merged.

The actual delta in this PR is tiny. The only thing new here is hooking up the `jj git init` command itself and inside `git_init` handling `--colocated` differently.

The rest of the PR is test and documentation.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (README.md, docs/, demos/)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/src/commands/git.rs
Comment (yuja) [inline cli/src/commands/git.rs:131]: Ah, good point. While it makes sense to not allow `jj git init --colocated` in existing git repo, it would be nice if we had an easy way to set up colocated repo in that way. `--git-repo=.` isn't useful if the destination path isn't cwd.

**Review by yuja — COMMENTED** [review]
> * I believe that --colocated doesn't need to go through the same
>   steps as --git-repo, after the repo is created. Is this correct?

Yes, but you'll need to set up `.jj/.gitignore`.

> * I've kept the current behaviour of `jj init  --git` by specifying
>   `colocated` false when invoked that way. I don't think it's worth
>   the effort  to add it in. Makes sense?

I don't think we'll need to add new feature to `jj init --git`.

File: cli/src/commands/git.rs
Comment (essiene) [reply cli/src/commands/git.rs:131]: I am thinking of following in subsequent PRs to add:

 --allow-existing - Use with --colocated to replace --git-repo=.; internally we'll still call init_git_external, but the user won't care.

--autopropagate? - Allow auto import/export to **existing** colocated repos. This will probably need wiring some flags to init_external_git.

If these seem like good ideas, I'll work on them after the initial simple version is merged.

Comment (yuja) [reply cli/src/commands/git.rs:131]: I'm not sure if adding flags is better, but this isn't a blocker of this PR. We can decide that later.

(I personally prefer `jj git init` that creates a colocated repo by default, and imports changes from `.git` repo if already exists.)

**Comment by essiene** [comment]
> > * I believe that --colocated doesn't need to go through the same
> >   steps as --git-repo, after the repo is created. Is this correct?
> 
> Yes, but you'll need to set up `.jj/.gitignore`.
Ahh... Thanks. Done.

> 
> > * I've kept the current behaviour of `jj init  --git` by specifying
> >   `colocated` false when invoked that way. I don't think it's worth
> >   the effort  to add it in. Makes sense?
> 
> I don't think we'll need to add new feature to `jj init --git`.
Thanks for confirming.

File: cli/src/commands/git.rs
Comment (essiene) [reply cli/src/commands/git.rs:131]: Makes sense. I'll bring this up for discussion on the issue itself.

Thanks.

File: cli/src/commands/git.rs
Comment (martinvonz) [inline cli/src/commands/git.rs:127]: nit: s/folder/directory/ (I feel like i hear the latter much more frequently, and you already use "directory" in other places in this text)

File: cli/src/commands/init.rs
Comment (martinvonz) [inline cli/src/commands/init.rs:38]: Since this is now deprecated, we should add `hide = true`, or what do you think? Same for `git_repo` below.

File: cli/src/commands/init.rs
Comment (martinvonz) [inline cli/src/commands/init.rs:60]: nit: I don't know if there's a difference between "colocated" and "collocated", but we seem to consistently spell it with one "l" everywhere else, so let's do that here too.

File: cli/src/commands/init.rs
Comment (PhilipMetzger) [reply cli/src/commands/init.rs:38]: Nit: I'd hide it and then print a warning message if we encounter it on the commandline, see #2747.

Comment (martinvonz) [reply cli/src/commands/init.rs:38]: Good point. That would also match what we're doing for `jj checkout` and `jj merge`.

File: cli/src/commands/git.rs
Comment (PhilipMetzger) [inline cli/src/commands/git.rs:438]: Nit: you can use a raw string `r#"Initialized repo in "{}""#`  here, which alleviates escaping.

File: cli/src/commands/git.rs
Comment (essiene) [reply cli/src/commands/git.rs]: good point. fixed.

File: cli/src/commands/git.rs
Comment (essiene) [reply cli/src/commands/git.rs]: Fixed. I have permanent identity crisis around folder/directory usage ;-)

File: cli/src/commands/init.rs
Comment (essiene) [reply cli/src/commands/init.rs]: Thanks. I missed that earlier. I've hidden both flags now.

File: cli/src/commands/git.rs
Comment (essiene) [reply cli/src/commands/git.rs]: Done. Thanks for the tip.

File: cli/src/commands/init.rs
Comment (essiene) [reply cli/src/commands/init.rs]: My bad. That's a typo on my part. Fixed.

File: cli/src/commands/init.rs
Comment (PhilipMetzger) [inline cli/src/commands/init.rs:66]: Optional: Could be "Use `jj git init` instead".

**Review by PhilipMetzger — APPROVED** [review]
LGTM. But I'll defer to martin or yuya for a final approval.

File: cli/src/commands/init.rs
Comment (essiene) [reply cli/src/commands/init.rs]: Makes sense. Done.

**Comment by essiene** [comment]
> LGTM. But I'll defer to martin or yuya for a final approval.

Thanks. I'll wait for a stamp from one of them.

File: cli/tests/test_git_init.rs
Comment (martinvonz) [inline cli/tests/test_git_init.rs:1]: 2024 :)

**Review by martinvonz — APPROVED** [review]

File: cli/tests/test_git_init.rs
Comment (essiene) [reply cli/tests/test_git_init.rs]: heh! Fixed :-D

**Pull request merged**
